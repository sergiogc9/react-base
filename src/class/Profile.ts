export interface ProfileObject {}

export class Profile {

	// public user: User;

	// constructor(profile: ProfileObject) {
	// 	this.user = new User(profile.user);
	// }
}
