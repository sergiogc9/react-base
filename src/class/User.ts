export class UserObject {
	public id: string;
	public name: string;

	constructor(values?: Partial<UserObject>) {
		if (values) Object.assign(this, values);
	}
}

export class User extends UserObject {

}
