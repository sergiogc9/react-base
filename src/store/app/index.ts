import { all } from 'redux-saga/effects';
import { combineReducers } from 'redux';

import * as auth from './auth';
import * as notifications from './notifications';
import * as profile from './profile';

// define sagas generator function to register all sagas
export function* sagas() {
	yield all([
		auth.sagas(),
		profile.sagas(),
		notifications.sagas()
	]);
}

export const reducers = combineReducers({
	auth: auth.reducers,
	profile: profile.reducers,
	notifications: notifications.reducers
});

export interface State {
	auth: auth.State;
	profile: profile.State;
	notifications: notifications.State;
}

export const INITIAL_STATE = {
	auth: auth.INITIAL_STATE,
	profile: profile.INITIAL_STATE,
	notifications: notifications.INITIAL_STATE
};
