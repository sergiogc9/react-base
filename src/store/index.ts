import { createBrowserHistory } from 'history';
import isEqual from 'lodash/isEqual';
import { combineReducers, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import { all } from 'redux-saga/effects';

import { connectRouter } from 'connected-react-router';

import { connectApi } from '@src/lib/ajax/Api';
import { getMiddleware, sagaMiddleware } from '@src/middleware';
import { State } from './types';

import * as app from './app';

const INITIAL_STATE: State = {
	app: app.INITIAL_STATE,
	router: { action: "POP", location: { hash: "", pathname: "/", search: "", state: undefined } }
};

// define sagas generator function to register all sagas
function* sagas() {
	yield all([
		app.sagas()
	]);
}

const history = createBrowserHistory();
// patch to avoid stacking same location multiple times in history
const historyPush = history.push;
let lastLocation = history.location;
history.listen(location => lastLocation = location);
history.push = (path: any, state: object = {}): void => {
	const lastPath = lastLocation.pathname + lastLocation.hash + lastLocation.search;
	if (lastLocation === null || path !== lastPath || !isEqual(state, lastLocation.state)) historyPush(path, state);
};

// create reducers

const reducers = combineReducers({
	app: app.reducers,
	router: connectRouter(history)
});

const store = createStore(reducers, composeWithDevTools(getMiddleware(history)));
connectApi(store);

sagaMiddleware.run(sagas);

export { INITIAL_STATE, sagas, history, reducers, store };
