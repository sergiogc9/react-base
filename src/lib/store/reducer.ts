import { CaseReducer, createSlice, EntityState, PayloadAction } from '@reduxjs/toolkit';

import { EntitiesFetchData } from 'store/entities';

// General reducer stuff
type CaseReducerWithPrepare<State, Payload> = {
	reducer: CaseReducer<State, {
		payload: Payload;
		type: string;
	}>;
	prepare: (payload: Payload) => {
		payload: Payload;
		meta: object | undefined;
	};
}

const getReducer = <State, Payload = void>(reducer: CaseReducer<State, PayloadAction<Payload>>, meta?: object): CaseReducerWithPrepare<State, Payload> => ({
	reducer,
	prepare: (payload: Payload) => ({ payload, meta })
});

// Api specific reducer stuff
export type DefaultPayloadError = { code: string, message: string };

type ApiOperatorOptions = {
	reload?: boolean // Used to reload page when optimistic api call fails
};

type ApiReducers<State, StartPayload, SuccessPayload, ErrorPayload> = {
	start?: CaseReducer<State, PayloadAction<StartPayload>>
	success?: CaseReducer<State, PayloadAction<SuccessPayload>>
	error?: CaseReducer<State, PayloadAction<ErrorPayload>>
}

const getApiReducers = <State, StartPayload = void, SuccessPayload = void, ErrorPayload = DefaultPayloadError>(
	reducers: ApiReducers<State, StartPayload, SuccessPayload, ErrorPayload>,
	options?: ApiOperatorOptions
): [
		CaseReducerWithPrepare<State, StartPayload>,
		CaseReducerWithPrepare<State, SuccessPayload>,
		CaseReducerWithPrepare<State, ErrorPayload>
	] => {
	const defaultReducer: CaseReducer<State> = state => state as State;
	return [
		getReducer<State, StartPayload>(reducers.start || defaultReducer, { api: 'start' }),
		getReducer<State, SuccessPayload>(reducers.success || defaultReducer, { api: 'success' }),
		getReducer<State, ErrorPayload>(reducers.error || defaultReducer, { api: 'error', reload: options?.reload })
	];
};

const getEntityFetchReducers = <T, StartPayload = void, SuccessPayload = void, ErrorPayload = DefaultPayloadError>() => {
	return getApiReducers<EntityState<T> & EntitiesFetchData, StartPayload, SuccessPayload, ErrorPayload>({
		start: state => { state.status = 'loading'; },
		success: state => { state.status = 'loaded'; },
		error: state => { state.status = 'error'; }
	});
};

export { getReducer, getApiReducers, getEntityFetchReducers, createSlice };
