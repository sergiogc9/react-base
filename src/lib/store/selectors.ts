import { useSelector as reduxUseSelector } from 'react-redux';

import { State } from 'store/types';

type Selector<T> = (state :State) => T;

const useSelector = <T>(selector: Selector<T>) => reduxUseSelector<State, T>(selector);

export { useSelector };
