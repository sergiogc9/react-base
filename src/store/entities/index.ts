import { combineReducers } from '@reduxjs/toolkit';

import selectors from './selectors';
import * as author from './Author';
import * as book from './Book';

export const reducer = combineReducers({
	author: author.reducer,
	book: book.reducer
});

export interface State {
	author: typeof author.INITIAL_STATE;
	book: typeof book.INITIAL_STATE;
}

export const INITIAL_STATE = {
	author: author.INITIAL_STATE,
	book: book.INITIAL_STATE
};

export { selectors };
