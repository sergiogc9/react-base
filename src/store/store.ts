import { combineReducers, configureStore } from '@reduxjs/toolkit';

import { customMiddlewares, sagaMiddleware } from 'middleware';
import * as auth from './auth';
import * as entities from './entities';
import * as notifications from './notifications';
import * as ui from './ui';
import sagas from './sagas';

// create reducers
const reducers = combineReducers({
	auth: auth.reducer,
	entities: entities.reducer,
	notifications: notifications.reducer,
	ui: ui.reducer
});

const store = configureStore({
	reducer: reducers,
	middleware: getDefaultMiddleware => getDefaultMiddleware().concat(...customMiddlewares)
});
const INITIAL_STATE = store.getState();

sagaMiddleware.run(sagas);

export { reducers, store, INITIAL_STATE };
