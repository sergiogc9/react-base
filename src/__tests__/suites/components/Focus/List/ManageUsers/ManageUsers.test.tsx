import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import CircularProgress from "react-md/lib/Progress/CircularProgress";
import SelectionControl from "react-md/lib/SelectionControls/SelectionControl";
import merge from "lodash/merge";

import { INITIAL_STATE } from "@src/store";
import { FocusObject, FocusFeeds } from "@src/class/Focus";
import { TenantObject, TenantUser } from "@src/class/Tenant";
import FocusListManageUsers from "@src/components/Focus/List/ManageUsers";
import TestHelper from '@src/__tests__/utils/Helper/TestHelper';

const mockStore = configureMockStore();

function getWrappedComponent(component: JSX.Element, stateSlice: object) {
	const store = mockStore(merge({}, INITIAL_STATE, stateSlice));
	return mount(<Provider store={store}>{component}</Provider>);
}

const document = TestHelper.getDocument();

const user1: TenantUser = {
	id: "fake-user-id-1",
	email: "fake@mail.com",
	first_name: "Fake",
	last_name: "Name",
	inserted_at: new Date(),
	updated_at: new Date(),
	role: "viewer"
};

const user2: TenantUser = {
	...user1,
	id: "fake-user-id-2",
	role: "clipper"
};

const tenantUsers: TenantUser[] = [user1, user2];

const focus: FocusObject = {
	deleted_at: new Date(1549287550563),
	feeds: { online: [], socialmedia: [], print: [] },
	id: "focus-id-1",
	inserted_at: new Date(1549287550563),
	name: "focus-name-1",
	oldest_document: document,
	updated_at: new Date(1549287550563),
	url_logo: "http://fake-url.com",
	acl_users: [
		{
			user_id: user1.id,
			discover_role: user1.role,
			last_name: user1.last_name,
			first_name: user1.first_name
		}
	]
};

const focus2: FocusObject = {
	...focus,
	id: "focus-id-2",
	name: "focus-name-2"
};

const focusList: FocusObject[] = [focus, focus2];

const tenant: TenantObject = {
	id: 'rd-girona-test',
	guid: '00034972-0000-0000-0000-000000000000',
	name: 'rd.girona.test',
	tier_properties: {
		name: 'custom',
		results: {
			online: true,
			social: true
		}
	},
	print_only: false,
	facebook_linked_ids: [
		'157193954975917'
	],
	settings: {
		categorization_mode: 'no_flc',
		currency: 'USD',
		display_influencers: true,
		facebook_url: 'https://www.facebook.com/conjunt.chapo',
		valuation_metric: 'miv'
	},
	users: tenantUsers
};

let onCloseMock: jest.Mock;

describe("FocusList", () => {
	let wrapper: any;

	beforeEach(() => {
		onCloseMock = jest.fn();
		wrapper = getWrappedComponent(<FocusListManageUsers
			focusId={focus.id}
			onClose={onCloseMock} />,
		{
			app: { profile: { tenant } },
			focus: {
				list: {
					focusList
				}
			}
		});
	});

	it("Manage focus users snapshot", () => {
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Manage focus users loading snapshot", () => {
		const component = getWrappedComponent(<FocusListManageUsers
			focusId={focus.id}
			onClose={onCloseMock} />,
		{
			app: { profile: { tenant } },
			focus: {
				list: {
					focusList,
					manageFocusLoadingUsers: [user1.id]
				}
			}
		});

		expect(component.html()).toMatchSnapshot();
		expect(component.find(CircularProgress)).toHaveLength(1);
	});

	it("Manage focus users toggle user status", () => {
		const component = getWrappedComponent(<FocusListManageUsers
			focusId={focus.id}
			onClose={onCloseMock} />,
		{
			app: { profile: { tenant: { ...tenant, users: [user1] } } },
			focus: {
				list: {
					focusList
				}
			}
		});

		component.find(SelectionControl).at(0).props().onChange(false);
		expect(component.find('div.md-switch-track--off')).toHaveLength(0);
	});

	it("Manage focus users close dialog", () => {
		wrapper.find(`button#launchmetricsDialogAcceptBtn`).simulate('click');
		expect(onCloseMock).toHaveBeenCalledTimes(1);
	});
});
