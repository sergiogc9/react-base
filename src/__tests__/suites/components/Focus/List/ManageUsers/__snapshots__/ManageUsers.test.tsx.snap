// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FocusList Manage focus users loading snapshot 1`] = `"<span class=\\"md-dialog-container md-overlay md-overlay--active md-pointer--hover launchmetrics-dialog\\" tabindex=\\"-1\\"><div aria-labelledby=\\"launchmetrics-dialog\\" id=\\"focusListManageFocusDialog\\" role=\\"dialog\\" class=\\"md-paper md-paper--5 md-dialog md-dialog--centered md-background--card\\"><h2 id=\\"focusListManageFocusDialog-title\\" class=\\"md-title md-title--dialog\\"><div class=\\"dialog-close\\"></div><div class=\\"dialog-header md-grid\\"><div class=\\"dialog-title\\">focus.manage_focus_title</div></div></h2><section class=\\"md-dialog-content md-dialog-content--padded\\"><span aria-hidden=\\"true\\"></span><div class=\\"dialog-container\\"><div class=\\"dialog-content\\"><div class=\\"manage-users-list\\"><div><div class=\\"manage-users-list-logo\\"><i class=\\"md-icon material-icons\\">account_circle</i></div><div class=\\"manage-users-list-info\\"><span class=\\"manage-users-list-info-name\\">Fake Name</span><span class=\\"manage-users-list-info-email\\">fake@mail.com</span></div><div class=\\"manage-users-list-toggle\\"><svg id=\\"focusListManageUsersToggleUserfake-user-id-1Loading\\" role=\\"progressbar\\" aria-valuemin=\\"0\\" aria-valuemax=\\"100\\" class=\\"md-progress md-progress--circular md-block-centered md-progress--circular-indeterminate\\" width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 66 66\\"><circle class=\\"md-circular-progress-path md-circular-progress-path--animated\\" stroke-width=\\"6\\" stroke-linecap=\\"round\\" cx=\\"33\\" cy=\\"33\\" r=\\"30\\"></circle></svg></div></div><div><div class=\\"manage-users-list-logo\\"><i class=\\"md-icon material-icons\\">account_circle</i></div><div class=\\"manage-users-list-info\\"><span class=\\"manage-users-list-info-name\\">Fake Name</span><span class=\\"manage-users-list-info-email\\">fake@mail.com</span></div><div class=\\"manage-users-list-toggle\\"><div class=\\"md-selection-control-container md-switch-container\\"><input id=\\"focusListManageUsersToggleUserfake-user-id-2Switch\\" type=\\"checkbox\\" class=\\"md-selection-control-input\\" name=\\"focusListManageUsersToggleUserfake-user-id-2Switch\\" aria-hidden=\\"true\\" value=\\"\\"><label for=\\"focusListManageUsersToggleUserfake-user-id-2Switch\\" class=\\"md-selection-control-label md-pointer--hover md-text\\"><div class=\\"md-switch-track md-pointer--hover md-switch-track--off\\"><div aria-labelledby=\\"focusListManageUsersToggleUserfake-user-id-2Switch-label\\" role=\\"checkbox\\" aria-checked=\\"false\\" class=\\"md-fake-btn md-pointer--hover md-fake-btn--no-outline md-switch-thumb md-switch-thumb--off\\" tabindex=\\"0\\" aria-pressed=\\"false\\"><div class=\\"md-ink-container md-ink-container--2x\\"></div></div></div><span id=\\"focusListManageUsersToggleUserfake-user-id-2Switch-label\\">focus.user_not_allowed</span></label></div></div></div></div></div></div></section><footer class=\\"md-dialog-footer md-dialog-footer--stacked dialog-footer md-grid center-buttons\\"><button id=\\"launchmetricsDialogAcceptBtn\\" type=\\"button\\" class=\\"md-btn md-btn--flat md-btn--text md-pointer--hover md-text md-inline-block md-btn--dialog btn--green btn--no-background\\"><div class=\\"md-ink-container\\"></div>focus.manage_focus_users_ok</button></footer></div></span>"`;

exports[`FocusList Manage focus users snapshot 1`] = `"<span class=\\"md-dialog-container md-overlay md-overlay--active md-pointer--hover launchmetrics-dialog\\" tabindex=\\"-1\\"><div aria-labelledby=\\"launchmetrics-dialog\\" id=\\"focusListManageFocusDialog\\" role=\\"dialog\\" class=\\"md-paper md-paper--5 md-dialog md-dialog--centered md-background--card\\"><h2 id=\\"focusListManageFocusDialog-title\\" class=\\"md-title md-title--dialog\\"><div class=\\"dialog-close\\"></div><div class=\\"dialog-header md-grid\\"><div class=\\"dialog-title\\">focus.manage_focus_title</div></div></h2><section class=\\"md-dialog-content md-dialog-content--padded\\"><span aria-hidden=\\"true\\"></span><div class=\\"dialog-container\\"><div class=\\"dialog-content\\"><div class=\\"manage-users-list\\"><div><div class=\\"manage-users-list-logo\\"><i class=\\"md-icon material-icons\\">account_circle</i></div><div class=\\"manage-users-list-info\\"><span class=\\"manage-users-list-info-name\\">Fake Name</span><span class=\\"manage-users-list-info-email\\">fake@mail.com</span></div><div class=\\"manage-users-list-toggle\\"><div class=\\"md-selection-control-container md-switch-container\\"><input id=\\"focusListManageUsersToggleUserfake-user-id-1Switch\\" type=\\"checkbox\\" class=\\"md-selection-control-input\\" name=\\"focusListManageUsersToggleUserfake-user-id-1Switch\\" aria-hidden=\\"true\\" value=\\"\\" checked=\\"\\"><label for=\\"focusListManageUsersToggleUserfake-user-id-1Switch\\" class=\\"md-selection-control-label md-pointer--hover md-text\\"><div class=\\"md-switch-track md-pointer--hover md-switch-track--on\\"><div aria-labelledby=\\"focusListManageUsersToggleUserfake-user-id-1Switch-label\\" role=\\"checkbox\\" aria-checked=\\"true\\" class=\\"md-fake-btn md-pointer--hover md-fake-btn--no-outline md-switch-thumb md-switch-thumb--on\\" tabindex=\\"0\\" aria-pressed=\\"false\\"><div class=\\"md-ink-container md-ink-container--2x\\"></div></div></div><span id=\\"focusListManageUsersToggleUserfake-user-id-1Switch-label\\">focus.user_allowed</span></label></div></div></div><div><div class=\\"manage-users-list-logo\\"><i class=\\"md-icon material-icons\\">account_circle</i></div><div class=\\"manage-users-list-info\\"><span class=\\"manage-users-list-info-name\\">Fake Name</span><span class=\\"manage-users-list-info-email\\">fake@mail.com</span></div><div class=\\"manage-users-list-toggle\\"><div class=\\"md-selection-control-container md-switch-container\\"><input id=\\"focusListManageUsersToggleUserfake-user-id-2Switch\\" type=\\"checkbox\\" class=\\"md-selection-control-input\\" name=\\"focusListManageUsersToggleUserfake-user-id-2Switch\\" aria-hidden=\\"true\\" value=\\"\\"><label for=\\"focusListManageUsersToggleUserfake-user-id-2Switch\\" class=\\"md-selection-control-label md-pointer--hover md-text\\"><div class=\\"md-switch-track md-pointer--hover md-switch-track--off\\"><div aria-labelledby=\\"focusListManageUsersToggleUserfake-user-id-2Switch-label\\" role=\\"checkbox\\" aria-checked=\\"false\\" class=\\"md-fake-btn md-pointer--hover md-fake-btn--no-outline md-switch-thumb md-switch-thumb--off\\" tabindex=\\"0\\" aria-pressed=\\"false\\"><div class=\\"md-ink-container md-ink-container--2x\\"></div></div></div><span id=\\"focusListManageUsersToggleUserfake-user-id-2Switch-label\\">focus.user_not_allowed</span></label></div></div></div></div></div></div></section><footer class=\\"md-dialog-footer md-dialog-footer--stacked dialog-footer md-grid center-buttons\\"><button id=\\"launchmetricsDialogAcceptBtn\\" type=\\"button\\" class=\\"md-btn md-btn--flat md-btn--text md-pointer--hover md-text md-inline-block md-btn--dialog btn--green btn--no-background\\"><div class=\\"md-ink-container\\"></div>focus.manage_focus_users_ok</button></footer></div></span>"`;
